File: DS-Algos/prime.py
from math import sqrt


def sieve(n):
    sieve_arr = [1] * (n + 1)  # Initialize with all elements as 1
    sieve_arr[0] = sieve_arr[1] = 0  # Mark 0 and 1 as non-prime

    for i in range(2, int(n**0.5) + 1):  # Optimized loop for efficiency
        if sieve_arr[i]:
            for j in range(i * i, n + 1, i):
                sieve_arr[j] = 0

    primes = [i for i in range(2, n + 1) if sieve_arr[i]]  # Collect primes
    return primes


def segmented_sieve(low, high, primes):
    low = max(low, 2)  # Handle cases of 0 and 1
    is_prime = [1] * (high - low + 1)
    if low > high:
        return []
    for p in primes:
        s = (low // p) * p
        if s < low:
            s += p
        if s == p:
            s += p

        for i in range(s, high + 1, p):
            is_prime[i - low] = 0

    ans = [i + low for i, prime in enumerate(is_prime) if prime]
    return ans


def extended_gcd(A, B):
    if B == 0:
        return (1, 0, A)
    xd, yd, gcd = extended_gcd(B, A % B)
    x = yd
    y = xd - (A // B) * yd
    return (x, y, gcd)


def modulo_inv(A, mod):
    ans = extended_gcd(A, mod)
    x = ans[0]
    return (x + mod) % mod


def diophantine(A, B, C):
    x, y, gcd = extended_gcd(A, B)
    if C % gcd != 0:
        return (-1, -1)
    return (x * C // gcd, y * C // gcd)


if __name__ == "__main__":
    # low, high = [int(i) for i in input().split()]
    # primes = sieve(int(sqrt(high) + 1))
    # N = 1e5 + 1
    # primes = sieve(N)
    # ans = segmented_sieve(low, high, primes)
    # print(*ans)
    print(modulo_inv(9, 10**9 + 5), (7 * modulo_inv(9, 10**9 + 5)) % (10**9 + 7))

File: DS-Algos/prime.cpp
#include <bits/stdc++.h>
#define vi vector<int>
#define tii tuple<int, int, int>
#define pii pair<int, int>
#define int long long int
using namespace std;
int n = 1e5 + 1;
vi sieve(int n);
vi segmentedSieve(int low, int high, vi &primes);
tii extended_gcd(int A, int B);
int modulo_inv(int A, int mod);
pii diophantine(int A, int B, int C);
vi primes = sieve(n);
int32_t main()
{
    int t = 1;
    // cin >> t;
    while (t--)
    {
        int low, high;
        cin >> low >> high;
        // vi ans = segmentedSieve(low, high, primes);
        tii ans = extended_gcd(low, high);
        cout << get<0>(ans) << " " << get<1>(ans) << " " << get<2>(ans) << endl;
        cout << "Modulo inverse of " << low << " is " << modulo_inv(low, 22) << endl;
        cout << "Modulo inverse of " << high << " is " << modulo_inv(high, 22) << endl;
        auto p = diophantine(low, high, 40);
        cout << "Solution " << p.first << " " << p.second << endl;
    }
    cout << endl;
    return 0;
}

vi sieve(int n)
{
    vi sieve_arr = vi(n + 1, 1);
    sieve_arr[0] = sieve_arr[1] = 0;
    for (int i = 2; i <= n; i++)
    {
        if (!sieve_arr[i])
        {
            for (int p = i * i; p <= n; p += i)
            {
                sieve_arr[p] = 1;
            }
        }
    }
    // return sieve_arr;
    vi primes;
    for (int i = 2; i <= n; i++)
    {
        if (sieve_arr[i])
            primes.push_back(i);
    }
    return primes;
}

vi segmentedSieve(int low, int high, vi &primes)
{
    // handling case of 0 and 1
    low = max(low, (int)2);
    if (low > high)
    {
        return {};
    }
    vi isPrime(high - low + 1, 1);
    vi ans;
    for (int p : primes)
    {
        int s = (low / p) * p;
        if (s < low)
            s += p;
        if (s == p)
            s += p;

        for (int i = s; i <= high; i += p)
        {
            isPrime[i - low] = 0;
        }
    }

    for (int i = 0; i < isPrime.size(); i++)
    {
        if (isPrime[i])
        {
            ans.push_back(i + low);
        }
    }
    return ans;
}

tii extended_gcd(int A, int B)
{
    if (B == 0)
        return {1, 0, A};
    auto [xd, yd, gcd] = extended_gcd(B, A % B);
    int x = yd;
    int y = xd - A / B * yd;
    return {x, y, gcd};
}

int modulo_inv(int A, int mod)
{
    tii ans = extended_gcd(A, mod);
    int x = get<0>(ans);
    return (x + mod) % mod;
}

pii diophantine(int A, int B, int C)
{
    auto [x, y, gcd] = extended_gcd(A, B);
    if (C % gcd != 0)
        return make_pair(-1, -1);
    return make_pair(x * C / gcd, y * C / gcd);
}